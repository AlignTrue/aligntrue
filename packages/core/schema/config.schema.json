{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://aligntrue.com/schema/config-v1.schema.json",
  "title": "AlignTrue Config v1",
  "description": "Configuration schema for AlignTrue CLI (solo/team/enterprise modes)",
  "type": "object",
  "required": [],
  "additionalProperties": true,
  "properties": {
    "version": {
      "type": "string",
      "description": "Config version identifier (e.g., '1' or semver '1.0.0')"
    },
    "mode": {
      "type": "string",
      "enum": ["solo", "team", "enterprise"],
      "description": "Operating mode: solo (default, no lockfile), team (lockfile+bundle), enterprise (all features)"
    },
    "modules": {
      "type": "object",
      "description": "Module feature flags (opt-in per module)",
      "additionalProperties": false,
      "properties": {
        "lockfile": {
          "type": "boolean",
          "description": "Enable lockfile generation for reproducibility (auto-enabled in team mode)"
        },
        "bundle": {
          "type": "boolean",
          "description": "Enable bundle generation for multi-source merging (auto-enabled in team mode)"
        },
        "checks": {
          "type": "boolean",
          "description": "Enable machine-checkable rules engine (default: true)"
        },
        "mcp": {
          "type": "boolean",
          "description": "Enable MCP server for editor integration (Phase 2+)"
        }
      }
    },
    "lockfile": {
      "type": "object",
      "description": "Lockfile-specific configuration",
      "additionalProperties": false,
      "properties": {
        "mode": {
          "type": "string",
          "enum": ["off", "soft", "strict"],
          "description": "Lockfile validation mode: off (no validation), soft (warn on mismatch), strict (error on mismatch)"
        }
      }
    },
    "git": {
      "type": "object",
      "description": "Git integration settings",
      "additionalProperties": false,
      "properties": {
        "mode": {
          "type": "string",
          "enum": ["ignore", "commit", "branch"],
          "description": "Git integration mode: ignore (no tracking), commit (auto-commit), branch (dedicated branch)"
        },
        "per_adapter": {
          "type": "object",
          "description": "Override git mode per exporter adapter",
          "additionalProperties": {
            "type": "string",
            "enum": ["ignore", "commit", "branch"]
          }
        }
      }
    },
    "sync": {
      "type": "object",
      "description": "Sync behavior configuration",
      "additionalProperties": false,
      "properties": {
        "auto_pull": {
          "type": "boolean",
          "description": "Automatically pull changes from primary agent on sync (default: true for solo, false for team)"
        },
        "primary_agent": {
          "type": "string",
          "description": "Primary agent to auto-pull from (e.g., 'cursor', 'copilot')"
        },
        "on_conflict": {
          "type": "string",
          "enum": ["prompt", "keep_ir", "accept_agent"],
          "description": "Conflict resolution strategy: prompt (interactive), keep_ir (default), accept_agent (auto-accept)"
        }
      }
    },
    "sources": {
      "type": "array",
      "description": "Source providers for rules (local files, catalog, git repos, URLs)",
      "items": {
        "type": "object",
        "required": ["type"],
        "additionalProperties": false,
        "properties": {
          "type": {
            "type": "string",
            "enum": ["local", "catalog", "git", "url"],
            "description": "Source provider type"
          },
          "path": {
            "type": "string",
            "description": "Local file path (required for type: local)"
          },
          "url": {
            "type": "string",
            "description": "Remote URL (required for type: git or url)"
          },
          "id": {
            "type": "string",
            "description": "Catalog pack ID (required for type: catalog)"
          },
          "version": {
            "type": "string",
            "description": "Version constraint for catalog packs (e.g., '^1.0.0')"
          }
        }
      }
    },
    "exporters": {
      "type": "array",
      "description": "List of exporter adapters to generate (e.g., 'cursor', 'agents-md')",
      "items": {
        "type": "string",
        "minLength": 1
      }
    },
    "scopes": {
      "type": "array",
      "description": "Hierarchical scopes for monorepo path-based rule application",
      "items": {
        "type": "object",
        "required": ["path"],
        "additionalProperties": false,
        "properties": {
          "path": {
            "type": "string",
            "description": "Relative path from workspace root (e.g., 'packages/frontend')"
          },
          "include": {
            "type": "array",
            "description": "Glob patterns to include within this scope",
            "items": {
              "type": "string"
            }
          },
          "exclude": {
            "type": "array",
            "description": "Glob patterns to exclude from this scope",
            "items": {
              "type": "string"
            }
          },
          "rulesets": {
            "type": "array",
            "description": "Rule IDs or pack references to apply in this scope",
            "items": {
              "type": "string"
            }
          }
        }
      }
    },
    "merge": {
      "type": "object",
      "description": "Merge strategy configuration for multi-source rules",
      "additionalProperties": false,
      "properties": {
        "strategy": {
          "type": "string",
          "enum": ["deep"],
          "description": "Merge strategy: deep (property-level merge by rule ID)"
        },
        "order": {
          "type": "array",
          "description": "Merge order (precedence): root < path < local",
          "items": {
            "type": "string",
            "enum": ["root", "path", "local"]
          },
          "uniqueItems": true
        }
      }
    },
    "performance": {
      "type": "object",
      "description": "Performance guardrails and resource limits",
      "additionalProperties": false,
      "properties": {
        "max_file_size_mb": {
          "type": "number",
          "minimum": 1,
          "maximum": 1000,
          "description": "Maximum file size in MB for source files (default: 10MB)"
        },
        "max_directory_depth": {
          "type": "number",
          "minimum": 1,
          "maximum": 50,
          "description": "Maximum directory depth for scanning (default: 10 levels)"
        },
        "ignore_patterns": {
          "type": "array",
          "items": {
            "type": "string"
          },
          "description": "Additional ignore patterns beyond .gitignore (glob format)"
        }
      }
    },
    "export": {
      "type": "object",
      "description": "Export behavior configuration for mode hints and token caps",
      "additionalProperties": true,
      "properties": {
        "mode_hints": {
          "type": "object",
          "description": "Control how execution mode metadata is exported to non-Cursor formats",
          "additionalProperties": false,
          "properties": {
            "default": {
              "type": "string",
              "enum": ["off", "metadata_only", "hints", "native"],
              "description": "Default mode hints behavior: off (no markers), metadata_only (JSON markers only), hints (markers + visible intent), native (exporter-specific)"
            },
            "overrides": {
              "type": "object",
              "description": "Per-exporter overrides for mode hints behavior",
              "additionalProperties": {
                "type": "string",
                "enum": ["off", "metadata_only", "hints", "native"]
              }
            }
          }
        },
        "max_hint_blocks": {
          "type": "number",
          "minimum": 1,
          "description": "Maximum number of rules to include mode hints for (default: 20)"
        },
        "max_hint_tokens": {
          "type": "number",
          "minimum": 100,
          "description": "Maximum token budget for mode hints across all rules (default: 1600)"
        }
      }
    },
    "backup": {
      "type": "object",
      "description": "Backup and restore configuration",
      "additionalProperties": false,
      "properties": {
        "auto_backup": {
          "type": "boolean",
          "description": "Automatically create backup before destructive operations (default: false)"
        },
        "keep_count": {
          "type": "number",
          "description": "Number of backups to keep (older backups auto-deleted, default: 10)",
          "minimum": 1,
          "maximum": 100
        },
        "backup_on": {
          "type": "array",
          "description": "Commands that trigger auto-backup (default: ['sync'])",
          "items": {
            "type": "string",
            "enum": ["sync", "restore", "import"]
          }
        }
      }
    },
    "overlays": {
      "type": "object",
      "description": "Overlay configuration for declarative customization of upstream rules (Phase 3.5)",
      "additionalProperties": false,
      "properties": {
        "overrides": {
          "type": "array",
          "description": "List of overlay definitions to apply to upstream rules",
          "items": {
            "type": "object",
            "required": ["selector"],
            "additionalProperties": false,
            "properties": {
              "selector": {
                "type": "string",
                "description": "Deterministic selector targeting exactly one rule (e.g., 'rule[id=...]')",
                "minLength": 1
              },
              "set": {
                "type": "object",
                "description": "Key-value pairs to set (supports nested paths with dot notation)",
                "additionalProperties": true
              },
              "remove": {
                "type": "array",
                "description": "List of keys to remove (supports nested paths with dot notation)",
                "items": {
                  "type": "string",
                  "minLength": 1
                }
              }
            }
          }
        },
        "limits": {
          "type": "object",
          "description": "Size limits for overlay configuration",
          "additionalProperties": false,
          "properties": {
            "max_overrides": {
              "type": "number",
              "minimum": 1,
              "maximum": 200,
              "description": "Maximum number of overrides allowed (default: 50)"
            },
            "max_operations_per_override": {
              "type": "number",
              "minimum": 1,
              "maximum": 100,
              "description": "Maximum combined set+remove operations per override (default: 20)"
            }
          }
        }
      }
    }
  }
}
