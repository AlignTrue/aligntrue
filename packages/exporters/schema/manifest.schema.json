{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://aligntrue.ai/schemas/adapter-manifest.json",
  "title": "AlignTrue Adapter Manifest",
  "description": "Describes an AlignTrue exporter adapter.",
  "type": "object",
  "required": ["name", "version", "description", "outputs"],
  "properties": {
    "name": {
      "type": "string",
      "pattern": "^[a-z0-9-]+$",
      "description": "Adapter name (lowercase alphanumeric with hyphens)"
    },
    "version": {
      "type": "string",
      "pattern": "^\\d+\\.\\d+\\.\\d+$",
      "description": "Semantic version (e.g., 1.0.0)"
    },
    "description": {
      "type": "string",
      "minLength": 10,
      "description": "Human-readable description of the adapter"
    },
    "outputs": {
      "type": "array",
      "items": {
        "type": "string"
      },
      "minItems": 1,
      "description": "File patterns produced by this adapter (e.g., .cursor/rules/*.mdc)"
    },
    "handler": {
      "type": "string",
      "description": "Optional: relative path to TypeScript handler (e.g., ./index.ts)"
    },
    "license": {
      "type": "string",
      "default": "MIT",
      "description": "License identifier (e.g., MIT, Apache-2.0)"
    },
    "fidelityNotes": {
      "type": "array",
      "items": {
        "type": "string"
      },
      "description": "Optional: list of semantic mapping limitations or caveats"
    }
  },
  "additionalProperties": false
}
