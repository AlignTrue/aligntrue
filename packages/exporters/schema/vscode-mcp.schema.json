{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://aligntrue.ai/schemas/vscode-mcp-v1.json",
  "title": "VS Code MCP Configuration v1",
  "description": "Defines the structure for a VS Code Multi-Project Configuration file generated by AlignTrue.",
  "type": "object",
  "required": ["version", "rules"],
  "properties": {
    "version": {
      "type": "string",
      "const": "v1",
      "description": "Format version marker for future evolution"
    },
    "generated_by": {
      "type": "string",
      "description": "Tool that generated this configuration"
    },
    "content_hash": {
      "type": "string",
      "pattern": "^[a-f0-9]{64}$",
      "description": "SHA-256 hash of canonical IR for drift detection"
    },
    "rules": {
      "type": "array",
      "description": "Array of rule objects with extracted vendor.vscode metadata",
      "items": {
        "type": "object",
        "required": ["id", "severity", "guidance"],
        "properties": {
          "id": {
            "type": "string",
            "description": "Unique rule identifier"
          },
          "severity": {
            "type": "string",
            "enum": ["error", "warn", "info"],
            "description": "Rule severity level"
          },
          "guidance": {
            "type": "string",
            "description": "Human-readable guidance text"
          },
          "scope": {
            "type": "string",
            "description": "Scope path this rule applies to"
          },
          "applies_to": {
            "type": "array",
            "items": {
              "type": "string"
            },
            "description": "File patterns this rule applies to"
          }
        },
        "additionalProperties": true,
        "description": "Rule object with vendor.vscode fields flattened to top level"
      }
    },
    "fidelity_notes": {
      "type": "array",
      "items": {
        "type": "string"
      },
      "description": "Notes about unmapped or vendor-specific fields"
    }
  },
  "additionalProperties": false
}
