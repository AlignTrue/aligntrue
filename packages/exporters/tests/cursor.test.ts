/**
 * Cursor exporter tests with snapshot validation
 */

import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import { readFileSync, rmSync, existsSync, mkdirSync } from 'fs'
import { join } from 'path'
import { CursorExporter, generateMdcFooter } from '../src/cursor/index.js'
import type { ScopedExportRequest, ExportOptions, ResolvedScope } from '../src/types.js'
import type { AlignRule } from '@aligntrue/schema'
import { parseYamlToJson } from '@aligntrue/schema'

const FIXTURES_DIR = join(import.meta.dirname, 'fixtures', 'cursor')
const TEST_OUTPUT_DIR = join(import.meta.dirname, 'temp-cursor-test-output')

describe('CursorExporter', () => {
  let exporter: CursorExporter

  beforeEach(() => {
    exporter = new CursorExporter()
    // Clean up test output directory
    if (existsSync(TEST_OUTPUT_DIR)) {
      rmSync(TEST_OUTPUT_DIR, { recursive: true })
    }
    mkdirSync(TEST_OUTPUT_DIR, { recursive: true })
  })

  afterEach(() => {
    // Clean up test output directory
    if (existsSync(TEST_OUTPUT_DIR)) {
      rmSync(TEST_OUTPUT_DIR, { recursive: true })
    }
  })

  describe('Plugin Interface', () => {
    it('implements ExporterPlugin interface', () => {
      expect(exporter.name).toBe('cursor')
      expect(exporter.version).toBe('1.0.0')
      expect(typeof exporter.export).toBe('function')
    })
  })

  describe('Basic Export', () => {
    it('exports single rule to .mdc file', async () => {
      const fixture = loadFixture('single-rule.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      expect(result.success).toBe(true)
      expect(result.filesWritten).toHaveLength(1)
      expect(result.filesWritten[0]).toMatch(/\.cursor\/rules\/aligntrue\.mdc$/)
      expect(result.contentHash).toMatch(/^[a-f0-9]{64}$/)
      expect(existsSync(result.filesWritten[0])).toBe(true)

      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toMatchSnapshot()
    })

    it('exports multiple rules to single .mdc file', async () => {
      const fixture = loadFixture('multiple-rules.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      expect(result.success).toBe(true)
      expect(result.filesWritten).toHaveLength(1)

      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('## Rule: testing.require-tests')
      expect(content).toContain('## Rule: docs.update-readme')
      expect(content).toContain('## Rule: style.use-prettier')
      expect(content).toMatchSnapshot()
    })

    it('throws error when rules array is empty', async () => {
      const request = createRequest([], createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      await expect(exporter.export(request, options)).rejects.toThrow(
        'CursorExporter requires at least one rule to export'
      )
    })
  })

  describe('Vendor.cursor Metadata Extraction', () => {
    it('extracts vendor.cursor fields to frontmatter', async () => {
      const fixture = loadFixture('with-vendor-cursor.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('cursor:')
      expect(content).toContain('testing.require-tests:')
      expect(content).toContain('ai_hint: "Suggest test scaffolding with vitest"')
      expect(content).toMatchSnapshot()
    })

    it('handles mixed vendor namespaces', async () => {
      const fixture = loadFixture('mixed-vendor.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      const content = readFileSync(result.filesWritten[0], 'utf-8')
      // Should extract cursor metadata
      expect(content).toContain('cursor:')
      expect(content).toContain('ai_hint: "Suggest vitest"')
      // Should include fidelity note about other agents
      expect(content).toContain('Vendor metadata for other agents preserved but not active: copilot, vscode')
      expect(content).toMatchSnapshot()
    })
  })

  describe('Fidelity Notes', () => {
    it('includes fidelity notes in footer', async () => {
      const fixture = loadFixture('multiple-rules.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      expect(result.fidelityNotes).toBeDefined()
      expect(result.fidelityNotes!.length).toBeGreaterThan(0)

      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('**Fidelity Notes:**')
      expect(content).toContain('applies_to patterns preserved in metadata but not enforced by Cursor')
    })

    it('tracks cross-agent vendor fields', async () => {
      const fixture = loadFixture('mixed-vendor.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      const note = result.fidelityNotes!.find(n => n.includes('other agents'))
      expect(note).toContain('copilot')
      expect(note).toContain('vscode')
    })
  })

  describe('Scope-to-Filename Mapping', () => {
    it('maps default scope to aligntrue.mdc', async () => {
      const fixture = loadFixture('single-rule.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      expect(result.filesWritten[0]).toContain('aligntrue.mdc')
    })

    it('maps named scope to scoped filename', async () => {
      const fixture = loadFixture('single-rule.yaml')
      const scope: ResolvedScope = {
        path: 'apps/web',
        normalizedPath: 'apps/web',
        isDefault: false,
        include: ['apps/web/**/*'],
      }
      const request = createRequest(fixture.rules, scope)
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      expect(result.filesWritten[0]).toContain('apps-web.mdc')
    })

    it('normalizes paths with slashes to hyphens', async () => {
      const fixture = loadFixture('single-rule.yaml')
      const scope: ResolvedScope = {
        path: 'packages/core/src',
        normalizedPath: 'packages/core/src',
        isDefault: false,
      }
      const request = createRequest(fixture.rules, scope)
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      expect(result.filesWritten[0]).toContain('packages-core-src.mdc')
    })
  })

  describe('Content Hash', () => {
    it('computes deterministic SHA-256 hash', async () => {
      const fixture = loadFixture('single-rule.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result1 = await exporter.export(request, options)
      
      // Clean up and export again
      rmSync(TEST_OUTPUT_DIR, { recursive: true })
      mkdirSync(TEST_OUTPUT_DIR, { recursive: true })
      
      const result2 = await exporter.export(request, options)

      expect(result1.contentHash).toBe(result2.contentHash)
      expect(result1.contentHash).toMatch(/^[a-f0-9]{64}$/)
    })

    it('includes content hash in footer', async () => {
      const fixture = loadFixture('single-rule.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('Content Hash:')
      expect(content).toContain(result.contentHash)
    })
  })

  describe('Dry Run Mode', () => {
    it('does not write files in dry run mode', async () => {
      const fixture = loadFixture('single-rule.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: true }

      const result = await exporter.export(request, options)

      expect(result.success).toBe(true)
      expect(result.filesWritten).toHaveLength(0)
      expect(result.contentHash).toMatch(/^[a-f0-9]{64}$/)
      expect(existsSync(join(TEST_OUTPUT_DIR, '.cursor', 'rules', 'aligntrue.mdc'))).toBe(false)
    })
  })

  describe('All Severities', () => {
    it('handles error, warn, and info severities', async () => {
      const fixture = loadFixture('all-severities.yaml')
      const request = createRequest(fixture.rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }

      const result = await exporter.export(request, options)

      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('**Severity:** error')
      expect(content).toContain('**Severity:** warn')
      expect(content).toContain('**Severity:** info')
      expect(content).toMatchSnapshot()
    })
  })

  describe('Cursor Mode Preservation', () => {
    it('exports mode: "always" as alwaysApply', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'error',
        applies_to: ['**/*'],
        mode: 'always',
        guidance: 'Test guidance',
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('alwaysApply: true')
    })

    it('exports mode: "intelligent" with description', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'warn',
        applies_to: ['**/*'],
        mode: 'intelligent',
        description: 'Intelligent mode active',
        guidance: 'Smart rule',
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('intelligent: true')
      expect(content).toContain('description: Intelligent mode active')
    })

    it('exports mode: "files" with applies_to as globs', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'info',
        applies_to: ['src/**/*.ts', 'tests/**/*.test.ts'],
        mode: 'files',
        guidance: 'Test guidance',
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('globs:')
      expect(content).toContain('- "src/**/*.ts"')
      expect(content).toContain('- "tests/**/*.test.ts"')
    })

    it('exports title and tags when present', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'error',
        applies_to: ['**/*'],
        mode: 'always',
        title: 'Test Rule',
        tags: ['typescript', 'quality'],
        guidance: 'Test guidance',
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('title: Test Rule')
      expect(content).toContain('tags:')
      expect(content).toContain('- typescript')
      expect(content).toContain('- quality')
    })

    it('restores unknown fields from vendor.cursor._unknown', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'error',
        applies_to: ['**/*'],
        mode: 'always',
        guidance: 'Test guidance',
        vendor: {
          cursor: {
            _unknown: {
              customField: 'custom value',
              futureFeature: 42
            }
          }
        }
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('customField: custom value')
      expect(content).toContain('futureFeature: 42')
    })

    it('separates schema fields and per-rule vendor metadata', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'error',
        applies_to: ['**/*'],
        mode: 'intelligent',
        description: 'File desc',
        guidance: 'Test guidance',
        vendor: {
          cursor: {
            ai_hint: 'Per-rule hint',  // Per-rule only
            quick_fix: true  // Per-rule only
          }
        }
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      // Schema fields in frontmatter (intelligent mode exports description)
      expect(content).toContain('intelligent: true')
      expect(content).toContain('description: File desc')
      // Per-rule vendor fields under cursor: key
      expect(content).toContain('cursor:')
      expect(content).toContain('test.rule.id:')
      expect(content).toContain('ai_hint: "Per-rule hint"')
      expect(content).toContain('quick_fix: true')
    })

    it('exports intelligent mode WITH globs (globs filter where AI relevance applies)', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'warn',
        applies_to: ['src/**/*.ts', 'lib/**/*.js'],
        mode: 'intelligent',
        description: 'Smart TypeScript/JS rule',
        guidance: 'AI decides relevance, but only on TypeScript and JavaScript files',
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      // Should have both intelligent mode AND globs
      expect(content).toContain('intelligent: true')
      expect(content).toContain('description: Smart TypeScript/JS rule')
      expect(content).toContain('globs:')
      expect(content).toContain('- "src/**/*.ts"')
      expect(content).toContain('- "lib/**/*.js"')
    })

    it('exports always mode WITH globs (globs filter which files get auto-applied)', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'error',
        applies_to: ['**/*.py', '**/*.pyi'],
        mode: 'always',
        guidance: 'Auto-apply only to Python files',
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      // Should have both always mode AND globs
      expect(content).toContain('alwaysApply: true')
      expect(content).toContain('globs:')
      expect(content).toContain('- "**/*.py"')
      expect(content).toContain('- "**/*.pyi"')
    })

    it('does not export globs when applies_to is default **/*', async () => {
      const rules: AlignRule[] = [{
        id: 'test.rule.id',
        severity: 'warn',
        applies_to: ['**/*'],  // Default - applies to all files
        mode: 'intelligent',
        description: 'Applies everywhere',
        guidance: 'No globs needed',
      }]
      
      const request = createRequest(rules, createDefaultScope())
      const options: ExportOptions = { outputDir: TEST_OUTPUT_DIR, dryRun: false }
      const result = await exporter.export(request, options)
      
      const content = readFileSync(result.filesWritten[0], 'utf-8')
      expect(content).toContain('intelligent: true')
      expect(content).toContain('description: Applies everywhere')
      // Should NOT have globs since it's the default
      expect(content).not.toContain('globs:')
    })
  })
})

describe('generateMdcFooter', () => {
  it('generates footer with content hash', () => {
    const hash = 'abc123'
    const notes: string[] = []

    const footer = generateMdcFooter(hash, notes)

    expect(footer).toContain('---')
    expect(footer).toContain('**Generated by AlignTrue**')
    expect(footer).toContain('Content Hash: abc123')
  })

  it('includes fidelity notes when provided', () => {
    const hash = 'abc123'
    const notes = ['Note 1', 'Note 2']

    const footer = generateMdcFooter(hash, notes)

    expect(footer).toContain('**Fidelity Notes:**')
    expect(footer).toContain('- Note 1')
    expect(footer).toContain('- Note 2')
  })

  it('omits fidelity notes section when empty', () => {
    const hash = 'abc123'
    const notes: string[] = []

    const footer = generateMdcFooter(hash, notes)

    expect(footer).not.toContain('**Fidelity Notes:**')
  })
})

// Helper functions

function loadFixture(filename: string): { rules: AlignRule[] } {
  const filepath = join(FIXTURES_DIR, filename)
  const yaml = readFileSync(filepath, 'utf-8')
  const data = parseYamlToJson(yaml) as any
  return { rules: data.rules }
}

function createRequest(rules: AlignRule[], scope: ResolvedScope): ScopedExportRequest {
  return {
    scope,
    rules,
    outputPath: join(TEST_OUTPUT_DIR, '.cursor', 'rules', 'aligntrue.mdc'),
  }
}

function createDefaultScope(): ResolvedScope {
  return {
    path: '.',
    normalizedPath: '.',
    isDefault: true,
    include: ['**/*'],
  }
}

