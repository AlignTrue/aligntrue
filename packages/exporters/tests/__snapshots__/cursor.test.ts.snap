// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CursorExporter > All Severities > handles error, warn, and info severities 1`] = `
"---
---
## Rule: critical.security

**Severity:** error

**Applies to:**
- \`**/*\`

Security violations must be fixed immediately.

## Rule: important.performance

**Severity:** warn

**Applies to:**
- \`**/*.ts\`

Performance issues should be addressed.

## Rule: nice-to-have.style

**Severity:** info

**Applies to:**
- \`**/*.ts\`

Style suggestions to consider.

---

**Generated by AlignTrue**
Content Hash: 8240511830b1fb8e49556a69760515943291f7882bbafda1d250669302a0ecfb

**Fidelity Notes:**
- applies_to patterns preserved in metadata but not enforced by Cursor
"
`;

exports[`CursorExporter > Basic Export > exports multiple rules to single .mdc file 1`] = `
"---
globs:
  - "**/*.ts"
---
## Rule: testing.require-tests

**Severity:** error

**Applies to:**
- \`**/*.ts\`

All features must have tests.

## Rule: docs.update-readme

**Severity:** warn

**Applies to:**
- \`README.md\`

Keep the README up to date with changes.

## Rule: style.use-prettier

**Severity:** info

**Applies to:**
- \`**/*.{ts,tsx,js,jsx}\`

Use Prettier for consistent formatting.

---

**Generated by AlignTrue**
Content Hash: 430f4e606cb24006cf0fab0deaef2412bc940504f1a2925b5ace04ce5f545f59

**Fidelity Notes:**
- applies_to patterns preserved in metadata but not enforced by Cursor
"
`;

exports[`CursorExporter > Basic Export > exports single rule to .mdc file 1`] = `
"---
globs:
  - "**/*.ts"
  - "**/*.tsx"
---
## Rule: testing.require-tests

**Severity:** warn

**Applies to:**
- \`**/*.ts\`
- \`**/*.tsx\`

All features must have tests to ensure reliability.
Write tests alongside your code, not as an afterthought.

---

**Generated by AlignTrue**
Content Hash: 7e39b22b4efdcbafb641f32942ee7d9f1060e765e552601c03cb4965efe2daa4

**Fidelity Notes:**
- applies_to patterns preserved in metadata but not enforced by Cursor
"
`;

exports[`CursorExporter > Vendor.cursor Metadata Extraction > extracts vendor.cursor fields to frontmatter 1`] = `
"---
globs:
  - "**/*.ts"
cursor:
  testing.require-tests:
    ai_hint: "Suggest test scaffolding with vitest"
    session_metadata: {"created":"2024-01-01"}
---
## Rule: testing.require-tests

**Severity:** warn

**Applies to:**
- \`**/*.ts\`

All features must have tests.

---

**Generated by AlignTrue**
Content Hash: ad9dd2c6dd7d874e4916ef796cdddd184d201c857a80541b70b63e41f674e7ef

**Fidelity Notes:**
- applies_to patterns preserved in metadata but not enforced by Cursor
"
`;

exports[`CursorExporter > Vendor.cursor Metadata Extraction > handles mixed vendor namespaces 1`] = `
"---
globs:
  - "**/*.ts"
cursor:
  testing.require-tests:
    ai_hint: "Suggest vitest"
---
## Rule: testing.require-tests

**Severity:** warn

**Applies to:**
- \`**/*.ts\`

All features must have tests.

---

**Generated by AlignTrue**
Content Hash: 362dc3efecaf82c6d8714dc5b27007102902a3541a3ebd89e13a4c7570c4eb40

**Fidelity Notes:**
- Vendor metadata for other agents preserved but not active: copilot, vscode
- applies_to patterns preserved in metadata but not enforced by Cursor
"
`;
